runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/ProcessRemoteBundleOperation.java

```diff
@@ -48,25 +61,98 @@
       new OutputReceiverFactory() {
         @Override
         public FnDataReceiver<?> create(String pCollectionId) {
-          return receivedElement -> {
-            LOG.debug("Consume element {}", receivedElement);
-            outputReceiverMap.get(pCollectionId).process((WindowedValue<?>) receivedElement);
-          };
+          return receivedElement -> receive(pCollectionId, receivedElement);
         }
       };
   private final StateRequestHandler stateRequestHandler;
   private final BundleProgressHandler progressHandler;
   private RemoteBundle remoteBundle;
+  private final DataflowExecutionContext<?> executionContext;
+  private final Map<String, ProcessBundleDescriptors.TimerSpec> timerOutputIdToSpecMap;
+  private final Map<String, Coder<BoundedWindow>> timerWindowCodersMap;
+  private final Map<String, ProcessBundleDescriptors.TimerSpec> timerIdToTimerSpecMap;
+  private final Map<String, Object> timerIdToKey;
+  private ExecutableStage executableStage;
 
   public ProcessRemoteBundleOperation(
-      OperationContext context,
+      ExecutableStage executableStage,
+      DataflowExecutionContext<?> executionContext,
+      DataflowOperationContext operationContext,
       StageBundleFactory stageBundleFactory,
       Map<String, OutputReceiver> outputReceiverMap) {
-    super(EMPTY_RECEIVER_ARRAY, context);
+    super(EMPTY_RECEIVER_ARRAY, operationContext);
+
     this.stageBundleFactory = stageBundleFactory;
-    this.outputReceiverMap = outputReceiverMap;
     this.stateRequestHandler = StateRequestHandler.unsupported();
     this.progressHandler = BundleProgressHandler.ignored();
+    this.executionContext = executionContext;
+    this.timerOutputIdToSpecMap = new HashMap<>();
+    this.timerWindowCodersMap = new HashMap<>();
+    this.executableStage = executableStage;
+    this.timerIdToKey = new HashMap<>();
+    this.outputReceiverMap = outputReceiverMap;
+    timerIdToTimerSpecMap = new HashMap<>();
+
+    for (Map<String, ProcessBundleDescriptors.TimerSpec> transformTimerMap :
+        stageBundleFactory.getProcessBundleDescriptor().getTimerSpecs().values()) {
+      for (ProcessBundleDescriptors.TimerSpec timerSpec : transformTimerMap.values()) {
+        timerIdToTimerSpecMap.put(timerSpec.timerId(), timerSpec);
+      }
+    }
+
+    for (Map<String, ProcessBundleDescriptors.TimerSpec> transformTimerMap :
+        stageBundleFactory.getProcessBundleDescriptor().getTimerSpecs().values()) {
+      for (ProcessBundleDescriptors.TimerSpec timerSpec : transformTimerMap.values()) {
+        timerOutputIdToSpecMap.put(timerSpec.outputCollectionId(), timerSpec);
+      }
+    }
+
+    for (RunnerApi.PTransform pTransform :
+        stageBundleFactory
+            .getProcessBundleDescriptor()
+            .getProcessBundleDescriptor()
+            .getTransformsMap()
+            .values()) {
+      for (String timerId : timerIdToTimerSpecMap.keySet()) {
+        if (!pTransform.getInputsMap().containsKey(timerId)) {
```

Comment:
I'd suggest to split this method into two: 
one for filling timerOutputIdToSpecMap
and other one for fillingtimerWindowCodersMap

Minimum request is to extract contents of this loop into separate method. It is too long and hard to track when reading. Giving it descriptive name (and probably method comment) will simplify understanding of what's going on.